yt/yt/core/concurrency/new_new_fair_share_thread_pool.cpp

```diff
@@ -0,0 +1,1368 @@
+                    YT_LOG_DEBUG_IF(VerboseLogging_, "Initial bucket excess time (Name: %v, ExcessTime: %v -> %v)",
+                        bucket->BucketName,
+                        bucket->ExcessTime,
+                        pool->LastBucketExcessTime);
+
+                    bucket->ExcessTime = pool->LastBucketExcessTime;
+                }
+
+                pool->ActiveBucketsHeap.Insert(bucket);
+                pool->BucketCounter.Record(pool->ActiveBucketsHeap.GetSize());
+                WaitHeap_.Insert(&bucket->EnqueuedTime);
+            }
+        });
+    }
+
+    void ServeBeginExecute(TThreadState* threadState, TCpuInstant currentInstant, TAction action)
+    {
+        VERIFY_SPINLOCK_AFFINITY(MainLock_);
+
+        YT_ASSERT(!threadState->Action.Callback);
+        YT_ASSERT(!threadState->Action.BucketHolder);
+
+        action.StartedAt = currentInstant;
+
+        threadState->AccountedAt = currentInstant;
+        threadState->Action = std::move(action);
+    }
+
+    void ServeEndExecute(TThreadState* threadState, TCpuInstant /*cpuInstant*/)
+    {
+        VERIFY_SPINLOCK_AFFINITY(MainLock_);
+
+        auto action = std::move(threadState->Action);
+        YT_ASSERT(!threadState->Action.Callback);
+        YT_ASSERT(!action.Callback);
+
+        if (!action.BucketHolder) {
+            // There was no action in begin execute.
+            return;
+        }
+
+        // Try not to change bucket ref count under lock.
+        auto bucket = std::move(action.BucketHolder);
+
+        auto& pool = *bucket->Pool;
+        YT_ASSERT(pool.PoolName == bucket->PoolName);
+
+        // LastActionsInQueue is used to update SizeCounter outside lock.
+        threadState->LastActionsInQueue = --pool.ActionCountInQueue;
+
+        // Do not destroy bucket pointer under lock. Move it in thread state in other place and
+        // destroy in corresponding thread after combiner.
+        threadState->BucketToUnref = std::move(bucket);
+    }
+
+    Y_NO_INLINE void UpdateExcessTime(TBucket* bucket, TCpuDuration duration, TCpuInstant currentInstant)
+    {
+        VERIFY_SPINLOCK_AFFINITY(MainLock_);
+
+        auto* pool = bucket->Pool.Get();
+
+        if (PoolWeightProvider_ && pool->NextUpdateWeightInstant < currentInstant) {
+            pool->NextUpdateWeightInstant = currentInstant + DurationToCpuDuration(TDuration::Seconds(1));
+            pool->InverseWeight = 1.0 / PoolWeightProvider_->GetWeight(pool->PoolName);
+        }
+
+        YT_LOG_DEBUG_IF(VerboseLogging_, "Increment excess time (BucketName: %v, PoolName: %v, ExcessTime: %v -> %v, Duration: %v, InverseWeight: %v)",
+            bucket->BucketName,
+            bucket->PoolName,
+            bucket->ExcessTime,
+            bucket->ExcessTime + duration * bucket->InverseWeight,
+            duration,
+            bucket->InverseWeight);
+
+        pool->ExcessTime += duration * pool->InverseWeight;
+        bucket->ExcessTime += duration * bucket->InverseWeight;
```

Comment:
I created a separate PR (to stable/23.2) to see a diff between `new_fair_share_thread_pool.cpp` and `new_new_fair_share_thread_pool.cpp`
https://github.com/Krock21/ytsaurus/pull/6/files

It does 3 things:
1. Introduces weights for buckets
2. Starts using new `IInvokerWithExpectedBytes`, that can pass expectedBytes
3. Introduces ExpectedBytes for actions, and changes logic for `ExcessTime`: 
  - Before it was using cpu time (`currentInstant - threadState.AccountedAt`), now it uses expectedBytes and resets them to 0 (to avoid counting them multiple times)
  - https://github.com/Krock21/ytsaurus/pull/6/files#diff-64af53abaeded1777ada927e39d6c877f3d1abd38d70554ce8a787bb8ce2d85aL997